pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,472 @@
+// pull intent into a pulled record and the original pull intent is deleted.
+//
+// This method is not thread-safe and it should only be called upon the creation
+// of the PullManager.
+func (f *PullManager) initialize(ctx context.Context) {
+	pullIntents, err := f.recordsAccessor.ListImagePullIntents()
+	if err != nil {
+		klog.ErrorS(err, "there was an error listing ImagePullIntents")
+		return
+	}
+
+	if len(pullIntents) == 0 {
+		return
+	}
+
+	images, err := f.imageService.ListImages(ctx)
+	if err != nil {
+		klog.ErrorS(err, "failed to list images")
+	}
+
+	inFlightPulls := sets.New[string]()
+	for _, intent := range pullIntents {
+		inFlightPulls.Insert(intent.Image)
+	}
+
+	// Each of the images known to the CRI might consist of multiple tags and digests,
+	// which is what we track in the ImagePullIntent - we need to go through all of these
+	// for each image.
+	for _, image := range images {
+		imageName := searchForExistingTagDigest(inFlightPulls, image)
+		if len(imageName) == 0 {
+			continue
+		}
+
+		if err := f.writePulledRecord(imageName, image.ID, nil); err != nil {
+			klog.ErrorS(err, "failed to write an image pull record", "imageRef", image.ID)
+			continue
+		}
+
+		if err := f.recordsAccessor.DeleteImagePullIntent(imageName); err != nil {
+			klog.V(2).InfoS("failed to remove image pull intent file", "imageName", imageName, "error", err)
+		}
+
+	}
+}
+
+// searchForExistingTagDigest loop through the `image` RepoDigests and RepoTags
+// and tries to find a digest/tag in `trackedImages`, which is a map of
+// containerImage -> pulling intent path.
+func searchForExistingTagDigest(trackedImages sets.Set[string], image kubecontainer.Image) string {
+	for _, digest := range image.RepoDigests {
+		if ok := trackedImages.Has(digest); ok {
+			return digest
+		}
+	}
+
+	for _, tag := range image.RepoTags {
+		if ok := trackedImages.Has(tag); ok {
+			return tag
+		}
+	}
+
+	return ""
+}
+
+type kubeSecretCoordinates struct {
+	UID       string
+	Namespace string
+	Name      string
+}
+
+// pulledRecordMergeNewCreds merges the credentials from `newCreds` into the `orig`
+// record for the `imageNoTagDigest` image.
+// `imageNoTagDigest` is the content of the `image` field from a pod's container
+// after any tag or digest were removed from it.
+func pulledRecordMergeNewCreds(orig *kubeletconfiginternal.ImagePulledRecord, imageNoTagDigest string, newCreds *kubeletconfiginternal.ImagePullCredentials) (*kubeletconfiginternal.ImagePulledRecord, bool) {
+	if newCreds == nil || (!newCreds.NodePodsAccessible && len(newCreds.KubernetesSecrets) == 0) {
+		return orig, false
+	}
+	ret := orig.DeepCopy()
+	selectedCreds, found := ret.CredentialMapping[imageNoTagDigest]
+	if !found {
+		if ret.CredentialMapping == nil {
+			ret.CredentialMapping = make(map[string]kubeletconfiginternal.ImagePullCredentials)
+		}
+		ret.CredentialMapping[imageNoTagDigest] = *newCreds
+	} else {
+		if newCreds.NodePodsAccessible {
```

Comment:
try to simplify this with an early check and return if the existing `selectedCreds` is already NodePodsAccessible

```
if selectedCred.NodePodsAccessible {
  // already accessible to all pods on the node
  return ret, false
}

if newCreds.NodePodsAccessible {
	selectedCreds.NodePodsAccessible = true
	selectedCreds.KubernetesSecrets = nil
	ret.LastUpdatedTime = metav1.Time{Time: time.Now()}
	return ret, true
}

// ... merge pull secrets
```